File: libs/moqtail-rs/src/model/control/control_message.rs
use crate::model::common::varint::BufVarIntExt;
use crate::model::error::ParseError;
use bytes::{Buf, Bytes};

use super::{
  client_setup::ClientSetup, constant::ControlMessageType, fetch::Fetch, fetch_cancel::FetchCancel,
  fetch_error::FetchError, fetch_ok::FetchOk, goaway::GoAway, max_request_id::MaxRequestId,
  publish_namespace::PublishNamespace, publish_namespace_cancel::PublishNamespaceCancel,
  publish_namespace_done::PublishNamespaceDone, publish_namespace_error::PublishNamespaceError,
  publish_namespace_ok::PublishNamespaceOk, requests_blocked::RequestsBlocked,
  server_setup::ServerSetup, subscribe::Subscribe, subscribe_done::SubscribeDone,
  subscribe_error::SubscribeError, subscribe_namespace::SubscribeNamespace,
  subscribe_namespace_error::SubscribeNamespaceError, subscribe_namespace_ok::SubscribeNamespaceOk,
  subscribe_ok::SubscribeOk, subscribe_update::SubscribeUpdate, track_status::TrackStatus,
  track_status_error::TrackStatusError, track_status_ok::TrackStatusOk, unsubscribe::Unsubscribe,
  unsubscribe_namespace::UnsubscribeNamespace,
};

#[derive(Debug, Clone, PartialEq)]
pub enum ControlMessage {
  PublishNamespace(Box<PublishNamespace>),
  PublishNamespaceCancel(Box<PublishNamespaceCancel>),
  PublishNamespaceError(Box<PublishNamespaceError>),
  PublishNamespaceOk(Box<PublishNamespaceOk>),
  ClientSetup(Box<ClientSetup>),
  Fetch(Box<Fetch>),
  FetchCancel(Box<FetchCancel>),
  FetchError(Box<FetchError>),
  FetchOk(Box<FetchOk>),
  Goaway(Box<GoAway>),
  MaxRequestId(Box<MaxRequestId>),
  ServerSetup(Box<ServerSetup>),
  Subscribe(Box<Subscribe>),
  SubscribeDone(Box<SubscribeDone>),
  SubscribeError(Box<SubscribeError>),
  SubscribeOk(Box<SubscribeOk>),
  SubscribeUpdate(Box<SubscribeUpdate>),
  RequestsBlocked(Box<RequestsBlocked>),
  TrackStatus(Box<TrackStatus>),
  TrackStatusOk(Box<TrackStatusOk>),
  TrackStatusError(Box<TrackStatusError>),
  PublishNamespaceDone(Box<PublishNamespaceDone>),
  Unsubscribe(Box<Unsubscribe>),
  SubscribeNamespace(Box<SubscribeNamespace>),
  SubscribeNamespaceOk(Box<SubscribeNamespaceOk>),
  SubscribeNamespaceError(Box<SubscribeNamespaceError>),
  UnsubscribeNamespace(Box<UnsubscribeNamespace>),
}

pub trait ControlMessageTrait: std::fmt::Debug {
  fn serialize(&self) -> Result<Bytes, ParseError>;
  fn parse_payload(payload: &mut Bytes) -> Result<Box<Self>, ParseError>
  where
    Self: Sized;
  fn get_type(&self) -> ControlMessageType;
}

impl ControlMessage {
  pub fn deserialize(bytes: &mut Bytes) -> Result<Self, ParseError> {
    let message_type = bytes.get_vi()?;
    let msg_type = ControlMessageType::try_from(message_type)?;

    if bytes.remaining() < 2 {
      return Err(ParseError::NotEnoughBytes {
        context: "ControlMessage::deserialize(payload_length)",
        needed: 2,
        available: 0,
      });
    }
    let payload_length = bytes.get_u16() as usize;

    if bytes.remaining() < payload_length {
      return Err(ParseError::NotEnoughBytes {
        context: "ControlMessage::deserialize(payload_length)",
        needed: payload_length,
        available: bytes.remaining(),
      });
    }

    let mut payload = bytes.copy_to_bytes(payload_length);
    let message = match msg_type {
      ControlMessageType::PublishNamespace => {
        PublishNamespace::parse_payload(&mut payload).map(ControlMessage::PublishNamespace)
      }
      ControlMessageType::PublishNamespaceCancel => {
        PublishNamespaceCancel::parse_payload(&mut payload)
          .map(ControlMessage::PublishNamespaceCancel)
      }
      ControlMessageType::PublishNamespaceDone => {
        PublishNamespaceDone::parse_payload(&mut payload).map(ControlMessage::PublishNamespaceDone)
      }
      ControlMessageType::PublishNamespaceError => {
        PublishNamespaceError::parse_payload(&mut payload)
          .map(ControlMessage::PublishNamespaceError)
      }
      ControlMessageType::PublishNamespaceOk => {
        PublishNamespaceOk::parse_payload(&mut payload).map(ControlMessage::PublishNamespaceOk)
      }
      ControlMessageType::ClientSetup => {
        ClientSetup::parse_payload(&mut payload).map(ControlMessage::ClientSetup)
      }
      ControlMessageType::Fetch => Fetch::parse_payload(&mut payload).map(ControlMessage::Fetch),
      ControlMessageType::FetchCancel => {
        FetchCancel::parse_payload(&mut payload).map(ControlMessage::FetchCancel)
      }
      ControlMessageType::FetchError => {
        FetchError::parse_payload(&mut payload).map(ControlMessage::FetchError)
      }
      ControlMessageType::FetchOk => {
        FetchOk::parse_payload(&mut payload).map(ControlMessage::FetchOk)
      }
      ControlMessageType::GoAway => GoAway::parse_payload(&mut payload).map(ControlMessage::Goaway),
      ControlMessageType::MaxRequestId => {
        MaxRequestId::parse_payload(&mut payload).map(ControlMessage::MaxRequestId)
      }
      ControlMessageType::ServerSetup => {
        ServerSetup::parse_payload(&mut payload).map(ControlMessage::ServerSetup)
      }
      ControlMessageType::Subscribe => {
        Subscribe::parse_payload(&mut payload).map(ControlMessage::Subscribe)
      }
      ControlMessageType::SubscribeDone => {
        SubscribeDone::parse_payload(&mut payload).map(ControlMessage::SubscribeDone)
      }
      ControlMessageType::SubscribeError => {
        SubscribeError::parse_payload(&mut payload).map(ControlMessage::SubscribeError)
      }
      ControlMessageType::SubscribeOk => {
        SubscribeOk::parse_payload(&mut payload).map(ControlMessage::SubscribeOk)
      }
      ControlMessageType::SubscribeUpdate => {
        SubscribeUpdate::parse_payload(&mut payload).map(ControlMessage::SubscribeUpdate)
      }
      ControlMessageType::RequestsBlocked => {
        RequestsBlocked::parse_payload(&mut payload).map(ControlMessage::RequestsBlocked)
      }
      ControlMessageType::TrackStatus => {
        TrackStatus::parse_payload(&mut payload).map(ControlMessage::TrackStatus)
      }
      ControlMessageType::TrackStatusOk => {
        TrackStatusOk::parse_payload(&mut payload).map(ControlMessage::TrackStatusOk)
      }
      ControlMessageType::TrackStatusError => {
        TrackStatusError::parse_payload(&mut payload).map(ControlMessage::TrackStatusError)
      }
      ControlMessageType::Unsubscribe => {
        Unsubscribe::parse_payload(&mut payload).map(ControlMessage::Unsubscribe)
      }
      ControlMessageType::SubscribeNamespace => {
        SubscribeNamespace::parse_payload(&mut payload).map(ControlMessage::SubscribeNamespace)
      }
      ControlMessageType::SubscribeNamespaceOk => {
        SubscribeNamespaceOk::parse_payload(&mut payload).map(ControlMessage::SubscribeNamespaceOk)
      }
      ControlMessageType::SubscribeNamespaceError => {
        SubscribeNamespaceError::parse_payload(&mut payload)
          .map(ControlMessage::SubscribeNamespaceError)
      }
      ControlMessageType::UnsubscribeNamespace => {
        UnsubscribeNamespace::parse_payload(&mut payload).map(ControlMessage::UnsubscribeNamespace)
      }
    }
    .map_err(|err| ParseError::ProtocolViolation {
      context: "ControlMessage::deserialize(payload)",
      details: err.to_string(),
    })?;

    if payload.has_remaining() {
      return Err(ParseError::ProtocolViolation {
        context: "ControlMessage::deserialize(final_check)",
        details: format!(
          "Extra {} bytes remaining in payload after parsing",
          payload.remaining()
        ),
      });
    };
    Ok(message)
  }

  pub fn serialize(&self) -> Result<Bytes, ParseError> {
    match self {
      ControlMessage::PublishNamespace(msg) => msg.serialize(),
      ControlMessage::PublishNamespaceCancel(msg) => msg.serialize(),
      ControlMessage::PublishNamespaceDone(msg) => msg.serialize(),
      ControlMessage::PublishNamespaceError(msg) => msg.serialize(),
      ControlMessage::PublishNamespaceOk(msg) => msg.serialize(),
      ControlMessage::ClientSetup(msg) => msg.serialize(),
      ControlMessage::Fetch(msg) => msg.serialize(),
      ControlMessage::FetchCancel(msg) => msg.serialize(),
      ControlMessage::FetchError(msg) => msg.serialize(),
      ControlMessage::FetchOk(msg) => msg.serialize(),
      ControlMessage::Goaway(msg) => msg.serialize(),
      ControlMessage::MaxRequestId(msg) => msg.serialize(),
      ControlMessage::ServerSetup(msg) => msg.serialize(),
      ControlMessage::Subscribe(msg) => msg.serialize(),
      ControlMessage::SubscribeDone(msg) => msg.serialize(),
      ControlMessage::SubscribeError(msg) => msg.serialize(),
      ControlMessage::SubscribeOk(msg) => msg.serialize(),
      ControlMessage::SubscribeUpdate(msg) => msg.serialize(),
      ControlMessage::RequestsBlocked(msg) => msg.serialize(),
      ControlMessage::TrackStatus(msg) => msg.serialize(),
      ControlMessage::TrackStatusOk(msg) => msg.serialize(),
      ControlMessage::TrackStatusError(msg) => msg.serialize(),
      ControlMessage::Unsubscribe(msg) => msg.serialize(),
      ControlMessage::SubscribeNamespace(msg) => msg.serialize(),
      ControlMessage::SubscribeNamespaceOk(msg) => msg.serialize(),
      ControlMessage::SubscribeNamespaceError(msg) => msg.serialize(),
      ControlMessage::UnsubscribeNamespace(msg) => msg.serialize(),
    }
  }

  /// Returns the message type of the control message.
  pub fn get_type(&self) -> ControlMessageType {
    match self {
      ControlMessage::PublishNamespace(_) => ControlMessageType::PublishNamespace,
      ControlMessage::PublishNamespaceCancel(_) => ControlMessageType::PublishNamespaceCancel,
      ControlMessage::PublishNamespaceDone(_) => ControlMessageType::PublishNamespaceDone,
      ControlMessage::PublishNamespaceError(_) => ControlMessageType::PublishNamespaceError,
      ControlMessage::PublishNamespaceOk(_) => ControlMessageType::PublishNamespaceOk,
      ControlMessage::ClientSetup(_) => ControlMessageType::ClientSetup,
      ControlMessage::Fetch(_) => ControlMessageType::Fetch,
      ControlMessage::FetchCancel(_) => ControlMessageType::FetchCancel,
      ControlMessage::FetchError(_) => ControlMessageType::FetchError,
      ControlMessage::FetchOk(_) => ControlMessageType::FetchOk,
      ControlMessage::Goaway(_) => ControlMessageType::GoAway,
      ControlMessage::MaxRequestId(_) => ControlMessageType::MaxRequestId,
      ControlMessage::ServerSetup(_) => ControlMessageType::ServerSetup,
      ControlMessage::Subscribe(_) => ControlMessageType::Subscribe,
      ControlMessage::SubscribeDone(_) => ControlMessageType::SubscribeDone,
      ControlMessage::SubscribeError(_) => ControlMessageType::SubscribeError,
      ControlMessage::SubscribeOk(_) => ControlMessageType::SubscribeOk,
      ControlMessage::SubscribeUpdate(_) => ControlMessageType::SubscribeUpdate,
      ControlMessage::RequestsBlocked(_) => ControlMessageType::RequestsBlocked,
      ControlMessage::TrackStatus(_) => ControlMessageType::TrackStatus,
      ControlMessage::TrackStatusOk(_) => ControlMessageType::TrackStatusOk,
      ControlMessage::TrackStatusError(_) => ControlMessageType::TrackStatusError,
      ControlMessage::Unsubscribe(_) => ControlMessageType::Unsubscribe,
      ControlMessage::SubscribeNamespace(_) => ControlMessageType::SubscribeNamespace,
      ControlMessage::SubscribeNamespaceOk(_) => ControlMessageType::SubscribeNamespaceOk,
      ControlMessage::SubscribeNamespaceError(_) => ControlMessageType::SubscribeNamespaceError,
      ControlMessage::UnsubscribeNamespace(_) => ControlMessageType::UnsubscribeNamespace,
    }
  }
}

#[cfg(test)]
mod tests {
  use crate::model::common::{pair::KeyValuePair, tuple::Tuple};

  use super::*;

  #[test]
  fn test_announce_roundtrip() {
    let request_id = 12345;
    let track_namespace = Tuple::from_utf8_path("god/dayyum");
    let parameters = vec![
      KeyValuePair::try_new_varint(0, 10).unwrap(),
      KeyValuePair::try_new_bytes(1, Bytes::from_static(b"wololoo")).unwrap(),
    ];
    let announce = PublishNamespace {
      request_id,
      track_namespace,
      parameters,
    };

    let mut buf = announce.serialize().unwrap();
    let deserialized = ControlMessage::deserialize(&mut buf).unwrap();
    if let ControlMessage::PublishNamespace(deserialized_announce) = deserialized {
      assert_eq!(*deserialized_announce, announce);
    } else {
      panic!("Expected ControlMessage::PublishNamespace variant");
    }
    assert!(!buf.has_remaining());
  }
}

File: libs/moqtail-ts/src/model/control/control_message.ts
/**
 * Copyright 2025 The MOQtail Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { FrozenByteBuffer } from '../common/byte_buffer'
import { ControlMessageType, controlMessageTypeFromBigInt, FetchType, GroupOrder } from './constant'
import { PublishNamespace } from './publish_namespace'
import { PublishNamespaceCancel } from './publish_namespace_cancel'
import { PublishNamespaceError } from './publish_namespace_error'
import { PublishNamespaceOk } from './publish_namespace_ok'
import { ClientSetup } from './client_setup'
import { Fetch } from './fetch'
import { FetchCancel } from './fetch_cancel'
import { FetchError } from './fetch_error'
import { FetchOk } from './fetch_ok'
import { GoAway } from './goaway'
import { MaxRequestId } from './max_request_id'
import { ServerSetup } from './server_setup'
import { Subscribe } from './subscribe'
import { SubscribeDone } from './subscribe_done'
import { SubscribeError } from './subscribe_error'
import { SubscribeOk } from './subscribe_ok'
import { SubscribeUpdate } from './subscribe_update'
import { RequestsBlocked } from './requests_blocked'
import { TrackStatus } from './track_status'
import { PublishNamespaceDone } from './publish_namespace_done'
import { Unsubscribe } from './unsubscribe'
import { SubscribeNamespace } from './subscribe_namespace'
import { SubscribeNamespaceOk } from './subscribe_namespace_ok'
import { SubscribeNamespaceError } from './subscribe_namespace_error'
import { UnsubscribeNamespace } from './unsubscribe_namespace'
import { NotEnoughBytesError } from '../error/error'
import { Tuple, KeyValuePair } from '../common'
import { TrackStatusOk } from './track_status_ok'
import { TrackStatusError } from './track_status_error'

export type ControlMessage =
  | PublishNamespace
  | PublishNamespaceCancel
  | PublishNamespaceError
  | PublishNamespaceOk
  | ClientSetup
  | Fetch
  | FetchCancel
  | FetchError
  | FetchOk
  | GoAway
  | MaxRequestId
  | ServerSetup
  | Subscribe
  | SubscribeDone
  | SubscribeError
  | SubscribeOk
  | SubscribeUpdate
  | RequestsBlocked
  | TrackStatus
  | TrackStatusOk
  | TrackStatusError
  | PublishNamespaceDone
  | Unsubscribe
  | SubscribeNamespace
  | SubscribeNamespaceOk
  | SubscribeNamespaceError
  | UnsubscribeNamespace

export namespace ControlMessage {
  export function deserialize(buf: FrozenByteBuffer): ControlMessage {
    const messageTypeRaw = buf.getVI()
    const messageType = controlMessageTypeFromBigInt(messageTypeRaw)
    const payloadLength = buf.getU16()
    if (buf.remaining < payloadLength)
      throw new NotEnoughBytesError('ControlMessage.deserialize(payload_bytes)', payloadLength, buf.remaining)
    const payloadBytes = buf.getBytes(payloadLength)
    const payload = new FrozenByteBuffer(payloadBytes)
    switch (messageType) {
      case ControlMessageType.PublishNamespace:
        return PublishNamespace.parsePayload(payload)
      case ControlMessageType.PublishNamespaceCancel:
        return PublishNamespaceCancel.parsePayload(payload)
      case ControlMessageType.PublishNamespaceError:
        return PublishNamespaceError.parsePayload(payload)
      case ControlMessageType.PublishNamespaceOk:
        return PublishNamespaceOk.parsePayload(payload)
      case ControlMessageType.ClientSetup:
        return ClientSetup.parsePayload(payload)
      case ControlMessageType.Fetch:
        return Fetch.parsePayload(payload)
      case ControlMessageType.FetchCancel:
        return FetchCancel.parsePayload(payload)
      case ControlMessageType.FetchError:
        return FetchError.parsePayload(payload)
      case ControlMessageType.FetchOk:
        return FetchOk.parsePayload(payload)
      case ControlMessageType.GoAway:
        return GoAway.parsePayload(payload)
      case ControlMessageType.MaxRequestId:
        return MaxRequestId.parsePayload(payload)
      case ControlMessageType.ServerSetup:
        return ServerSetup.parsePayload(payload)
      case ControlMessageType.Subscribe:
        return Subscribe.parsePayload(payload)
      case ControlMessageType.SubscribeDone:
        return SubscribeDone.parsePayload(payload)
      case ControlMessageType.SubscribeError:
        return SubscribeError.parsePayload(payload)
      case ControlMessageType.SubscribeOk:
        return SubscribeOk.parsePayload(payload)
      case ControlMessageType.SubscribeUpdate:
        return SubscribeUpdate.parsePayload(payload)
      case ControlMessageType.RequestsBlocked:
        return RequestsBlocked.parsePayload(payload)
      case ControlMessageType.TrackStatus:
        return TrackStatus.parsePayload(payload)
      case ControlMessageType.TrackStatus:
        return TrackStatus.parsePayload(payload)
      case ControlMessageType.PublishNamespaceDone:
        return PublishNamespaceDone.parsePayload(payload)
      case ControlMessageType.Unsubscribe:
        return Unsubscribe.parsePayload(payload)
      case ControlMessageType.SubscribeNamespace:
        return SubscribeNamespace.parsePayload(payload)
      case ControlMessageType.SubscribeNamespaceOk:
        return SubscribeNamespaceOk.parsePayload(payload)
      case ControlMessageType.SubscribeNamespaceError:
        return SubscribeNamespaceError.parsePayload(payload)
      case ControlMessageType.UnsubscribeNamespace:
        return UnsubscribeNamespace.parsePayload(payload)
      default:
        // This case should ideally be unreachable if controlMessageTypeFromBigInt is exhaustive
        // or throws on unknown types. If it can return a type not in the switch,
        // an error here is appropriate.
        throw new Error(`Unknown or unhandled ControlMessageType: ${messageType}`)
    }
  }

  export function serialize(msg: ControlMessage): FrozenByteBuffer {
    return msg.serialize()
  }
}

if (import.meta.vitest) {
  const { describe, test, expect } = import.meta.vitest

  describe('ControlMessage', () => {
    describe('PublishNamespace', () => {
      function buildTestPublishNamespace(): PublishNamespace {
        return new PublishNamespace(12345n, Tuple.fromUtf8Path('god/dayyum'), [
          KeyValuePair.tryNewVarInt(0, 10),
          KeyValuePair.tryNewBytes(1, new TextEncoder().encode('wololoo')),
        ])
      }

      test('should roundtrip PublishNamespace correctly', () => {
        const announce = buildTestPublishNamespace()
        const serialized = ControlMessage.serialize(announce)
        const deserialized = ControlMessage.deserialize(serialized)
        expect(deserialized).toEqual(announce)
      })

      test('should roundtrip PublishNamespace with excess trailing bytes', () => {
        const announce = buildTestPublishNamespace()
        const serialized = ControlMessage.serialize(announce).toUint8Array()
        const excessBytes = new Uint8Array(serialized.length + 3)
        excessBytes.set(serialized)
        excessBytes.set([9, 1, 1], serialized.length)

        const buf = new FrozenByteBuffer(excessBytes)
        const deserialized = ControlMessage.deserialize(buf)
        expect(deserialized).toEqual(announce)
        expect(buf.remaining).toBe(3) // Check that excess bytes are still there
        expect(Array.from(buf.getBytes(3))).toEqual([9, 1, 1])
      })

      test('should throw on partial PublishNamespace message', () => {
        const announce = buildTestPublishNamespace()
        const serialized = ControlMessage.serialize(announce).toUint8Array()
        const partial = serialized.slice(0, Math.floor(serialized.length / 2))
        const buf = new FrozenByteBuffer(partial)
        expect(() => ControlMessage.deserialize(buf)).toThrow(NotEnoughBytesError)
      })
    })

    describe('Fetch', () => {
      function buildTestFetch(): Fetch {
        const requestId = 161803n
        const subscriberPriority = 15
        const groupOrder = GroupOrder.Descending
        const joiningRequestId = 119n
        const joiningStart = 73n
        const type = FetchType.Relative
        const parameters = [
          KeyValuePair.tryNewVarInt(4444, 12321n),
          KeyValuePair.tryNewBytes(1, new TextEncoder().encode('fetch me ok')),
        ]
        return new Fetch(
          requestId,
          subscriberPriority,
          groupOrder,
          { type, props: { joiningRequestId, joiningStart } },
          parameters,
        )
      }

      test('should roundtrip Fetch correctly', () => {
        const fetchMsg = buildTestFetch()
        const serialized = ControlMessage.serialize(fetchMsg)
        const deserialized = ControlMessage.deserialize(serialized)
        expect(deserialized).toEqual(fetchMsg)
      })

      test('should roundtrip Fetch with excess trailing bytes', () => {
        const fetchMsg = buildTestFetch()
        const serialized = ControlMessage.serialize(fetchMsg).toUint8Array()
        const excessBytes = new Uint8Array(serialized.length + 3)
        excessBytes.set(serialized)
        excessBytes.set([8, 2, 2], serialized.length)

        const buf = new FrozenByteBuffer(excessBytes)
        const deserialized = ControlMessage.deserialize(buf)
        expect(deserialized).toEqual(fetchMsg)
        expect(buf.remaining).toBe(3)
        expect(Array.from(buf.getBytes(3))).toEqual([8, 2, 2])
      })

      test('should throw on partial Fetch message', () => {
        const fetchMsg = buildTestFetch()
        const serialized = ControlMessage.serialize(fetchMsg).toUint8Array()
        const partial = serialized.slice(0, Math.floor(serialized.length / 2))
        const buf = new FrozenByteBuffer(partial)
        expect(() => ControlMessage.deserialize(buf)).toThrow(NotEnoughBytesError)
      })
    })
  })
}
